### PR TITLE
Factor peek/poke out of simif_t

### PR DESCRIPTION
**Pull request by davidbiancolin**
I'd like to get to a point where we can build most of the driver source as a single library -- this as a stepping stone towards building a single driver. My plan is check in mmio address struct definitions alongside their bridges so that we can remove the macro guards, and eventually stop including the generated header in all compilation units.

This biggest blocker of this are derived simif classes that use the peek-poke interface, since:
- the mmio struct definition changes from target to target
- there's a lot of code that uses this IF 

We've long wanted to remove peek-poke from simif. However, peek poke is still useful for testing (but not for user's deployment) so here i pull out all peek-poke related methods out of simif into a new derived class used only in midasexamples/fasedtests. In a future PR i will set about removing the requirement that peek-poke bridge must be used in the target. 

#### UI / API Impact

Further slims down the god class `simif_t`. 
- Changes the semantics of finish to avoid diamond-pattern challenges with multiple inheritance
  - This method should be defined by class implementing the host abstraction layer (simif_f1/simif_emul). 

Most people don't write custom derived classes of simif_t that use peek poke unless they are writing tests. 

#### Verilog / AGFI Compatibility

No change.

### Contributor Checklist
- [x] Did you set dev as the base branch?
- [x] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [ ] Did you delete any extraneous prints/debugging code?
- [x] Did you state the UI / API impact?
- [x] Did you specify the Verilog / AGFI compatibility impact?

- [ ] (If applicable) Did you regenerate and publicly share default AGFIs?

- [ ] (If applicable) Did you mark the PR as "Please Backport"?
- [ ] (On merge) Did you update release notes in the dev-to-master PR ?

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

File: sim/midas/src/main/cc/simif.cc [hunk 1]
```diff
@@ -15,23 +15,19 @@ double diff_secs(midas_time_t end, midas_time_t start) {
...
   CLOCKBRIDGEMODULE_0_substruct_create;
   this->clock_bridge_mmio_addrs = CLOCKBRIDGEMODULE_0_substruct;
 }
 
-void simif_t::init(int argc, char** argv, bool log) {
+void simif_t::init(int argc, char** argv) {
+  // Do any post-constructor initialization required before requesting MMIO
+  this->host_init(argc, argv);
   while(!read(this->master_mmio_addrs->INIT_DONE));
-  this->log = log;
```
Comment (abejgonzalez) [inline sim/midas/src/main/cc/simif.cc:34]: How does this `log` boolean get enabled? This seems to be only used in the `simif_peek_poke` classes (at least from the diff)... but its a `private` member and thus this `log` isn't used?

File: sim/src/main/cc/fasedtests/fasedtests_top.cc [hunk 2]
```diff
@@ -111,11 +115,7 @@ void fasedtests_top_t::run() {
         zero_out_dram();
     }
     fprintf(stderr, "Commencing simulation.\n");
-    uint64_t start_hcycle = hcycle();
-    uint64_t start_time = timestamp();
-
-    // Assert reset T=0 -> 50
```
Comment (abejgonzalez) [inline sim/src/main/cc/fasedtests/fasedtests_top.cc:117]: This isn't needed because of the reset pulse bridge right?

File: sim/src/main/cc/fasedtests/test_harness_bridge.cc [hunk 3]
```diff
@@ -24,6 +24,10 @@ test_harness_bridge_t::test_harness_bridge_t(
 // it then reads uarch event counts from the FASED instance and compares them against
 // expected values
 void test_harness_bridge_t::tick(){
+  // Wait for reset to complete.
+  if (sim->actual_tcycle() < 100)
```
Comment (abejgonzalez) [inline sim/src/main/cc/fasedtests/test_harness_bridge.cc:28]: Why is this needed?

File: sim/src/main/cc/firesim/firesim_top.cc [hunk 4]
```diff
@@ -570,7 +570,7 @@ void firesim_top_t::run() {
 
     while (!simulation_complete() && !has_timed_out()) {
         run_scheduled_tasks();
-        step(get_largest_stepsize(), false);
+        take_steps(get_largest_stepsize(), false);
```
Comment (abejgonzalez) [inline sim/src/main/cc/firesim/firesim_top.cc:573]: So it looks like `step` was deleted and thus you replaced it with this. Can you elaborate a bit on what `step` did? I think my confusion is around what `t` means/meant (I know it moved to the `peek_poke` class).

**Review by abejgonzalez — COMMENTED** [review]
Most of this LGTM. I would just like to understand a couple points before the full stamp.

File: sim/src/main/cc/fasedtests/test_harness_bridge.cc (hunk 3 above)
Comment (davidbiancolin) [reply sim/src/main/cc/fasedtests/test_harness_bridge.cc:28]: There's no longer a blocking call to target_reset. This little hack does the job for these tests.

File: sim/src/main/cc/firesim/firesim_top.cc (hunk 4 above)
Comment (davidbiancolin) [reply sim/src/main/cc/firesim/firesim_top.cc:573]: t was essentially a cycle count, vis-a-vis steps taken by peek poke. If you used non-blocking peeks / pokes (rare) this would be pretty non-sensical. 

What `step` and it's underlying implementation `take_steps` does is write (indirectly) to the peek-poke widget some number of credits. When the credits are exhausted it stops producing tokens and the simulator eventually stalls. So "steps" here is in terms of cycles in whatever clock domain the peek-poke widget is in. 
We should probably change how time-control is handled to make more sense in a multi-clock, peek-poke-less simulator (in my dissertation branch the hub model will not advance past a certain time thats been plumbed out to MMIO), and rename this.

**Review by abejgonzalez — APPROVED** [review]
SGTM
